client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,179 @@
+defaults.driver = localforage[ defaults.driver ];
+
+/**
+ * SyncHandler class
+ */
+export class SyncHandler {
+	/**
+	 * Create a SyncHandler instance
+	 *
+	 * @param {Object} [setup] - sync setup
+	 * @param {Function} handler - wpcom handler
+	 *
+	 * @return {Function} sync-handler wrapper
+	 */
+	constructor( setup, handler ) {
+		if ( typeof setup === 'function' ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerInstance( handler );
+	}
+
+	syncHandlerInstance( handler ) {
+		return ( params, wrappedResponseCallback ) => {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return this.reqHandler( params, wrappedResponseCallback );
+			}
+
+			const clonedParams = Object.assign( {}, params );
+			const { path } = clonedParams;
+
+			// whitelist barrier
+			if ( ! isWhitelisted( params ) ) {
+				debug( 'not whitelisted: skip %o request', path );
+				return handler( params, wrappedResponseCallback );
+			};
+
+			// response sent flag
+			let responseSent = false;
+
+			// generate an unique resource key
+			const key = this.generateKey( params );
+
+			debug( 'starting to get resource ...' );
+
+			this.retrieveRecord( key, ( err, localRecord ) => {
+				if ( err ) {
+					// @TODO improve error handling here
+					console.error( err );
+				}
+
+				// let's be optimistic
+				if ( localRecord ) {
+					responseSent = true;
+
+					debug( '%o stored(%o). Let\'s be optimistic ...\n', path, localRecord );
+					wrappedResponseCallback( null, localRecord.body );
+				} else {
+					debug( 'No data for %o\n', path );
+				}
+
+				// * background sync process *
+				debug( 'requesting %o to WP.com', path );
+				handler( params, ( resErr, resData ) => {
+					if ( resErr ) {
+						if ( responseSent ) {
+							return;
+						}
+
+						return wrappedResponseCallback( resErr );
+					}
+
+					debug( 'WP.com response %o -> %o', path, resData );
+
+					const isPOSTRequest = clonedParams &&
+						clonedParams.method &&
+						'post' === clonedParams.method.toLowerCase();
+
+					if ( ! isPOSTRequest ) {
+						let storingData = {
+							__sync: {
+								key,
+								synced: new Date().toString(),
+								syncing: false
+							},
+							body: resData,
+							params: clonedParams
+						};
+
+						this.storeRecord( key, storingData, storedErr => {
+							if ( storedErr ) {
+								console.error( storedErr );
+								if ( ! responseSent ) {
+									wrappedResponseCallback( err, resData );
+								}
+							}
+
+							// extra response
+							let _resData = Object.assign(
+								{},
+								resData,
+								{ __syncResponse: true }
+							);
+
+							debug( 'background response sent' );
+							wrappedResponseCallback( err, _resData );
```

Comment:
IMHO we should run the response when we have the data. It doesn't seem necessary to wait until after we have saved the data into localforage before we run the callback.